### PR TITLE
Remove WebGPUEngine dependency on Engine

### PR DESCRIPTION
**Pull request by deltakosh**
Please bear with me folks ;)
It will be long and painful but at the end it should be great as WebGPU will not get ANY webgl related content ;)

Wish me luck!

Step 1: Make `WebGPUEngine` inherits from `ThinEngine`
Step 2: Move the hierarchy to be like this:

- AbstractEngine
  - WebgpuEngine
  - ThinEngine
     - Engine

This way we have full isolation. But I want to apologize to reviewers. This is a monster PR like none before :(

The only breaking change I'm anticipating is that every .engine properties across the lib will now be of `AbstractEngine` instead of `Engine` (but same methods so the impact should be almost non existent).

ThinEngine and Engine are almost unchanged. The big rework is two folds:
- Declare all specific methods (the one different between webgl and webgpu) as `abstract` methods on `AbstractEngine`
- Move Common functions from `ThinEngine` to `AbstractEngine`

With that in place, we will have a clear and easy way to implements `ThinWebGPUEngine`.

@RaananW: I would need your help as I moved NpmPackage and Version from `ThinEngine` to `AbstractEngine`.
Edit: I did it but want to be sure with your approval ;)

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14931/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14931/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14931/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14931/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14931/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14931/merge#BCU1XR#0

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

File: packages/tools/tests/test/visualization/webgpu_webgpu_log.txt
Comment (RaananW) [inline packages/tools/tests/test/visualization/webgpu_webgpu_log.txt:1]: not really needed :-) I must have missed this on the gitignore.

**Review by RaananW — COMMENTED** [review]
Before fully reviewing this amazing PR, a small note - there is one more place to update. To avoid build tools dependency when using free js scripts, the standalone script updateVersion.js (in the /scripts/ directory) needs an update as well.

**Review by sebavan — APPROVED** [review]

**Review by RaananW — COMMENTED** [review]
There is a breaking change of TypeScript's typing.

Once you change scene.getEngine to AbstractEngine, every typescript project that have used this function will need to cast to the right engine, otherwise it will fail. Same goes to all of the other getEngine calls.

A way to solve it is to infer the engine type using generics from the constructor, but this will require adding generics to all modules using an engine.

This is a good example of a typescript snippet that fails - https://playground.babylonjs.com/?snapshot=refs/pull/14931/merge#YX6IB8#674

**Comment by deltakosh** [comment]
I agree with that but I also consider it is a very small sized number of projects. 
As the problem only appears is TS it will not be a runtime issue but a build time issue that will be extremely easy for the user to fix

**Comment by deltakosh** [comment]
I mentioned that problem in my description:)

**Review by RaananW — APPROVED** [review]

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14931/merge/testResults/webgpuplaywright/index.html

**Pull request merged**